hazelcast/src/hazelcast/client/connection/ClientConnectionManagerImpl.cpp

```diff
@@ -0,0 +1,842 @@
+                            }
+
+                spi::impl::ClientExecutionServiceImpl::shutdownExecutor("cluster", *clusterConnectionExecutor, logger);
+                stopEventLoopGroup();
+                connectionListeners.clear();
+                heartbeat->shutdown();
+
+                connectionStrategy->shutdown();
+
+
+                if (alive.compareAndSet(true, false)) {
+                    return;
+                }
+
+                activeConnectionsFileDescriptors.clear();
+                activeConnections.clear();
+                socketConnections.clear();
+            }
+
+            boost::shared_ptr<Connection>
+            ClientConnectionManagerImpl::getOrConnect(const Address &address) {
+                return getOrConnect(address, false);
+            }
+
+            boost::shared_ptr<Connection> ClientConnectionManagerImpl::getOwnerConnection() {
+                boost::shared_ptr<Address> address = ownerConnectionAddress;
+                if (address.get() == NULL) {
+                    return boost::shared_ptr<Connection>();
+                }
+                boost::shared_ptr<Connection> connection = getActiveConnection(*address);
+                return connection;
+
+            }
+
+            boost::shared_ptr<Connection> ClientConnectionManagerImpl::connectAsOwner(const Address &address) {
+                boost::shared_ptr<Connection> connection;
+                try {
+                    logger.info() << "Trying to connect to " << address << " as owner member";
+                    connection = getOrConnect(address, true);
+                    client.onClusterConnect(connection);
+                    setOwnerConnectionAddress(boost::shared_ptr<Address>(new Address(connection->getRemoteEndpoint())));
+                    logger.info() << "Setting " << *connection << " as owner with principal " << *principal.get();
+                    fireConnectionEvent(LifecycleEvent::CLIENT_CONNECTED);
+                    connectionStrategy->onConnectToCluster();
+                } catch (exception::IException &e) {
+                    logger.warning() << "Exception during initial connection to " << address << ", exception " << e;
+                    if (NULL != connection.get()) {
+                        std::ostringstream reason;
+                        reason << "Could not connect to " << address << " as owner";
+                        connection->close(reason.str().c_str(), boost::shared_ptr<exception::IException>(e.clone()));
+                    }
+                    return boost::shared_ptr<Connection>();
+                }
+                return connection;
+            }
+
+            boost::shared_ptr<Connection>
+            ClientConnectionManagerImpl::createSocketConnection(const Address &address, bool ownerConnection) {
+                boost::shared_ptr<Connection> conn(
+                        new Connection(address, client, inSelector, outSelector, socketFactory, ownerConnection));
```

Comment:
Connection class should not get `ownerConnection` in constructor. It will be set after the authentication.

Reply:
Also createSocketConnection method with `ownerConnection` is not necessary.